Add conv2d transpose NNAPI converter

axitkhurana opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61025
* #61024
* #61023
* #61022
* #61021
* **#59529**
* #59434

Summary:

* Conv2d transpose support
* Quantize WIP

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28926335](https://our.internmc.facebook.com/intern/diff/D28926335)

Commit:
Conv transpose WIP

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28926335).

Commit:
Update on "Conv transpose WIP"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28926335](https://our.internmc.facebook.com/intern/diff/D28926335)

[ghstack-poisoned]

Commit:
Update on "Conv transpose WIP"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28926335](https://our.internmc.facebook.com/intern/diff/D28926335)

[ghstack-poisoned]

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28926335).

Commit:
Update on "Add conv2d transpose NNAPI converter"


Summary:

* Conv2d transpose support
* Quantize WIP

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28926335](https://our.internmc.facebook.com/intern/diff/D28926335)

[ghstack-poisoned]

Commit:
Update on "Add conv2d transpose NNAPI converter"


Summary:

* Conv2d transpose support
* Quantize WIP

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28926335](https://our.internmc.facebook.com/intern/diff/D28926335)

[ghstack-poisoned]

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28926335).

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28926335).

Commit:
Update on "Add conv2d transpose NNAPI converter"


Summary:

* Conv2d transpose support
* Quantize WIP

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28926335](https://our.internmc.facebook.com/intern/diff/D28926335)

[ghstack-poisoned]

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28926335).

Commit:
Update on "Add conv2d transpose NNAPI converter"


Summary:

* Conv2d transpose support
* Quantize WIP

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28926335](https://our.internmc.facebook.com/intern/diff/D28926335)

[ghstack-poisoned]

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28926335).

Commit:
Update on "Add conv2d transpose NNAPI converter"


Summary:

* Conv2d transpose support
* Quantize WIP

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28926335](https://our.internmc.facebook.com/intern/diff/D28926335)

[ghstack-poisoned]

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28926335).

Commit:
Update on "Add conv2d transpose NNAPI converter"


Summary:

* Conv2d transpose support
* Quantize WIP

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28926335](https://our.internmc.facebook.com/intern/diff/D28926335)

[ghstack-poisoned]

dreiss commented on torch/backends/_nnapi/serializer.py:
```diff
@@ -1658,10 +1658,13 @@ def add_qlinear(self, node):
         ctype, value = self.get_constant_value(jit_bias)
         if ctype.kind() == "NoneType":
-            nnapi_bias_tensor = torch.zeros(weight_tensor.size()[0], dtype=weight_tensor.dtype)
+            if transpose:
+                nnapi_bias_tensor = torch.zeros(weight_tensor.size()[1], dtype=weight_tensor.dtype)
+            else:
+                nnapi_bias_tensor = torch.zeros(weight_tensor.size()[0], dtype=weight_tensor.dtype)
```
```
bias_idx = 1 if transpose else 0
nnapi_bias_tensor = torch.zeros(weight_tensor.size()[bias_idx], dtype=weight_tensor.dtype)
```

dreiss commented on test/test_nnapi.py:
```diff
@@ -503,6 +503,62 @@ def test_conv2d(self):
+        convert_arg = torch.zeros(*convert_dims)
+        convert_args = dict(self.float_and_quant_and_nhwc(convert_arg, 0.03, 128))
+
+        class ConvTransposeMod(torch.nn.Module):
```
Can you use `torch.nn.ConvTranspose2d`?

dreiss commented on test/test_nnapi.py:
```diff
@@ -503,6 +503,62 @@ def test_conv2d(self):
                     )
 
+    def test_conv2d_transpose(self):
+        weight = torch.randn((128, 256, 2, 2))
+        input_dim = (30, 128, 7, 7)
```
Do we need the data to be this large?

axitkhurana replied:
Definitely not, was dealing with that size during conversion ended up using it in the test too

Commit:
Update on "Add conv2d transpose NNAPI converter"


Summary:

* Conv2d transpose support
* Quantize WIP

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28926335](https://our.internmc.facebook.com/intern/diff/D28926335)

[ghstack-poisoned]

Commit:
Update on "Add conv2d transpose NNAPI converter"


Summary:

* Conv2d transpose support
* Quantize WIP

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28926335](https://our.internmc.facebook.com/intern/diff/D28926335)

[ghstack-poisoned]

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28926335).

Commit:
Update on "Add conv2d transpose NNAPI converter"


Summary:

* Conv2d transpose support
* Quantize WIP

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28926335](https://our.internmc.facebook.com/intern/diff/D28926335)

[ghstack-poisoned]

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28926335).

Commit:
Update on "Add conv2d transpose NNAPI converter"


Summary:

* Conv2d transpose support
* Quantize WIP

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28926335](https://our.internmc.facebook.com/intern/diff/D28926335)

[ghstack-poisoned]

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28926335).
